### DISCUSSION TITLE
How to manage plugin config (for self-developed Google pubsub Plugins)

### DISCUSSION THREAD

**Discussion by pret-nitish** [emqx/emqx#14089] [Q&A]
Hi Team ,

We are trying to use google pubsub plugins emqx 5.8.1 version. but it is not working.

We have created plugins via custom plugins method. and we put the binary in plugins folder. same is visible in gui plugins window.

since we want to communicate to google pubsub   via emqx and publish messages to google pubsub topic from emqx. we are not able to see anywhere where we can put details and make communication like from connector.

Please let us know is there any other steps do we need to perform in emqx 5.8.1 version.

**Comment by zmstone** [comment]
Plugins do not support UI(dashboard) aid config management.


maybe consider using enterprise edition.

**Reply by pret-nitish** [reply]
Can you please share step by step setup to use plugins via cli config management.
if there is any link also we can refer please share the same.

**Comment by zmstone** [comment]
- A plugin can bootstrap itself with default configs in `priv/config.hocon`. see plugin dir layout here: https://docs.emqx.com/en/emqx/latest/extensions/plugins.html

- Config updated at runtime will be stored as `data/plugins/plugin_name/config.hocon` (documented in the link above)
 > Configuration updates made at runtime will be written in HOCON format to data/plugins/<PLUGIN_NAME>/config.hocon, and the old configuration file will be backed up each time the configuration is updated.

- To update config after plugin is installed, one can call this API.
https://docs.emqx.com/en/emqx/v5.8/admin/api-docs.html#tag/Plugins/paths/~1plugins~1%7Bname%7D~1config/put

- For enterprise edition you can add a avro schema for config validation, and annotations for avro schema to support config management from the dashboard UI.

**Reply by pret-nitish** [reply]
I am trying this method only but unable to make it work. can you please provide any further reference or link to make it work.

Please let me know if you need anything from my side.

**Reply by pret-nitish** [reply]
We checked this link where it has modified logic of on_client_subscribe in plugins file to be able to make plugin work.
https://www.emqx.com/en/blog/exploring-the-extensibility-of-emqx

We want also like whenever any message publish to xyz topic and same topic is present in gcp pubsub so we want that published message from emqx  to be send to gcp pubsub. we are new to this and dont have idea about erl.
